### DISCUSSION TITLE
Splitting Airflow pods between clusters (control and data planes)

### DISCUSSION THREAD

**Discussion by frizollo** [apache/airflow#27667] [Q&A] [answered]
Please, do you know how can we configure Airflow to run with this approach below?
This approach considers the use of two or more Kubernetes clusters.
>> One cluster is composed by Airflow web server + scheduler + etc pods, at Control Plane.
>> The other(s) cluster(s) is(are) composed by Airflow worker (Kubernetes pod operator), at Data Plane.  

Have you ever used Airflow in this way? Any clue of how to achieve it?

**Comment by potiuk** [comment]
Well. I am not sure what you mean by control plane and data plane in this context (what exactly do you mean by separatting airflow between them). Could you please explain what your requirements are? The names are pretty meaningless in this context without the explnation I am afraid.

**Comment by frizollo** [comment]
The requirements are: 
- To have a dedicated cluster to define the Control Plane, where a single and scallable Airflow service orchestrates DAGs to run tasks to process data from many other sources/clusters;
- The clusters where data resides defines the Data Plane level;
- Data Plane level can have many clusters (a cluster per geographic location or customer constraints);
- Data processing (where the worker pods are instantiated) should be performed at Data Plane level, to be compliant with data privacy and customer restriciton polocies;
- DAGs should be shcheduled and managed at Control Plane level, but the DAG tasks should be processed within each Data Plane cluster;
- DAG tasks are set to use Kubernetes Pods Operator, where each task instantiates a dedicated pod to process data. 

I hope the requirements above can help a litlle bit more. :) 

Do you have an idea of the beast practice to define these two layers (Control Plane, for orchestration; and Data Plane, for execution) and interconnect them?

![image](https://user-images.githubusercontent.com/99293131/201763108-0213c6dc-a59a-462d-9167-ba8ac71def33.png)

**Comment by potiuk** [comment, marked as answer]
Surely. You can do it even now but it's far more complex that the simple "control plane/data plane" separation.

KubernetesExecutor allow syou to set in_cluster = False and specify configuration to connect to another cluster to run actual PODs executing tasks.

Look at https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#in-cluster and whole https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#kubernetes section for all details.

Note that currently each of those K8S Task Pods in your Data Plane will need to have access to a central Metadata DB and DAG files, so they will not be fully isolated:

*  each of the PODS will have to talk to a DB 
*  have DAG folder mounted - either via shared file system or git sync or similar. 

See architecture here: https://airflow.apache.org/docs/apache-airflow/stable/executor/kubernetes.html

The DB access might be in the future - optionally - replaced with the Internal DB API server (AIP-44) https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-44+Airflow+Internal+API so the Task PODs will only have to have access to the Internal API server.

You can also use CeleryExecutor (or even Celery Kubernetes Executor) and each of the "Data Plane" PODs can also be a Celery Worker (but then it not only needs a MetaDB access but also access to a shared RabbitMQ/Redis queue).  

As of Airflow 2.3, there is also Triggerer which sits somewhere in between your Control and Data plane and is kinda both - both central control point but also run (via async.io) trigger code that belongs to individual tasks.

So your picture should be quite a bit more  complex if you want to make sure all involved components are included.

**Comment by potiuk** [comment]
Also - it's entirely up to you how you do the deployment in this case. You are free to do it in the way that is best for your case - whichever components you want to keep on whichever cluster is entirely up to you, you just need to remember that there are some parts (DB, Redis/RabbotMq, Internal API server in the future, etc.) are included in your deployment.
